### ISSUE TITLE
Assertion in EnsureIsReadOnlyAttributeExists during rename

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#24351]
**Version Used**: 

Private build.

```
---------------------------
Assertion Failed: Abort=Quit, Retry=Debug, Ignore=Continue
---------------------------




   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.EnsureIsReadOnlyAttributeExists(DiagnosticBag diagnostics, Location location, Boolean modifyCompilationForRefReadOnly) in C:\R0\src\Compilers\CSharp\Portable\Symbols\Compilation_WellKnownMembers.cs:line 465

   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.AfterMembersChecks(DiagnosticBag diagnostics) in C:\R0\src\Compilers\CSharp\Portable\Symbols\Source\SourceMemberContainerSymbol.cs:line 1419

   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken) in C:\R0\src\Compilers\CSharp\Portable\Symbols\Source\SourceMemberContainerSymbol.cs:line 507

   at Microsoft.CodeAnalysis.CSharp.Symbol.ForceCompleteMemberByLocation(SourceLocation locationOpt, Symbol member, CancellationToken cancellationToken) in C:\R0\src\Compilers\CSharp\Portable\Symbols\Symbol.cs:line 712

   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTy......

<truncated>
---------------------------
Abort   Retry   Ignore   
---------------------------
```

Crash dump:
%internal_share%\public\tomat\Bugs\Roslyn24351

**Comment by jcouv** [comment]
Tagging @OmarTawfik 
@tmat Do you have any details on the rename that caused this?

**Comment by tmat** [comment]
@jcouv Yes, i'd hope so. If not we need to look into how can we make the dumps more useful. I was in the middle of editing in Roslyn doing a bunch of changes. I don't know what exactly led to the assert.

**Comment by OmarTawfik** [comment]
Looking into it.

**Comment by OmarTawfik** [comment]
From what I can see, there was a rename code action executing, which renamed `ActiveInstructionId.ActiveMethodId` to `ActiveInstructionId.MethodId`, but then compilation assert failed on compiling the new code, because another type `Microsoft.CodeAnalysis.EditAndContinue.EditSession.Analysis` was a `readonly struct`, and it called `Compilation.EnsureIsReadOnlyAttributeExists()` after it was already frozen. I'll try to figure if I can dump the whole compilation to disk for a more consistent repro, but so far failing as I cannot call into any native methods.
cc @VSadov for the readonly structs.

**Comment by tmat** [comment]
@OmarTawfik You can perhaps try to repo in https://github.com/dotnet/roslyn/tree/ActiveStatements15.6.VS branch where these types are defined, if you can't find some kind of minimal repro involving readonly structs.

**Comment by OmarTawfik** [comment]
No luck reproducing. But I have a hypothesis. The IDE creates instances of `RecoverableSyntaxTree` during edit operations, and then adds them to the current compilation along with existing the compiler `ParsedSyntaxTree` instances. Is it possible that at some part during the pipeline of (`DocumentExtensions.GetErrorsAsync()`) these trees are mutable/mutated to the new rename value (instead of creating new syntax trees) breaking the immutability of the compilation object? @tmat @dotnet/roslyn-ide?

Other than that, and without a repro, I can see us moving that check earlier in the pipeline (binding, or symbol creation), and therefore never triggering it that late. But I prefer to rule out other possibilities first.

**Comment by jasonmalinowski** [comment]
@OmarTawfik That is always possible, but it's probably the same type of "possible" in that an IDE bug could always be the C# compiler compiling IDE code wrong. It's _possible_, but you might want to exclude more straightforward explanations first. :smile:

**Comment by OmarTawfik** [comment]
@jasonmalinowski what I'm trying to know is, during rename, does the IDE create new syntax trees (with the rename operation) and update the compilation (creating a new one effectively), or does it just update the contents of the existing syntax tree object, and then getting diagnostics again (effectively breaking immutability)?

**Comment by jasonmalinowski** [comment]
@OmarTawfik Oh I see your question. We most definitely don't break immutability -- that would be crazy. :smile: The "recoverable" indirection means that if we are under memory pressure we can toss them out of memory and page them back in. But that process should not be observable to you or any client.

**Comment by OmarTawfik** [comment]
Thanks for explaining :)

**Comment by AlekseyTs** [comment]
@OmarTawfik I might have found the reason for the assert. 
CSharpCompilation.FreezeDeclarationDiagnostics() sets _needsGeneratedAttributes_IsFrozen to true. It is called from CSharpCompilation.GetSourceDeclarationDiagnostics(...), which could actually force only partial completion for the source assembly if a syntax tree is passed in. However, it calls FreezeDeclarationDiagnostics regardless whether this is the case. So, I can imagine IDE requesting diagnostics from semantic model and causing the freeze to happen even though some types (from other syntax trees) haven't been completed yet. When we complete the types, the assert is triggered.

**Comment by OmarTawfik** [comment]
@AlekseyTs that's very helpful! We'll need to make a distinction between completing the full assembly (which we should freeze after), and a pass triggered by the IDE (which we shouldn't).
I'll try to repro and write some tests in the morning! Thanks :)

**Comment by OmarTawfik** [comment]
Fixed in #24489
